internal/checkers/http_const.go

```diff
@@ -0,0 +1,129 @@
+//	assert.HTTPBodyContains(t, handler, http.MethodGet, "/index", nil, "counter")
+type HTTPConst struct{}
+
+// NewHTTPConst constructs Regexp checker.
+func NewHTTPConst() HTTPConst  { return HTTPConst{} }
+func (HTTPConst) Name() string { return "http-const" }
+
+func (checker HTTPConst) Check(_ *analysis.Pass, call *CallMeta) *analysis.Diagnostic {
+	var suggestedFixes []analysis.SuggestedFix
+	switch call.Fn.NameFTrimmed {
+	case "HTTPBody",
+		"HTTPBodyContains",
+		"HTTPBodyNotContains",
+		"HTTPError",
+		"HTTPRedirect",
+		"HTTPSuccess":
+		if len(call.Args) < 2 {
+			return nil
+		}
+		suggestedFix := newHTTPMethodReplacement(call.Args[1])
+		if suggestedFix == nil {
+			return nil
+		}
+		suggestedFixes = append(suggestedFixes, *suggestedFix)
+	case "HTTPStatusCode":
+		if len(call.Args) < 5 {
+			return nil
+		}
+		if suggestedFix := newHTTPMethodReplacement(call.Args[1]); suggestedFix != nil {
+			suggestedFixes = append(suggestedFixes, *suggestedFix)
+		}
+		if suggestedFix := newHTTPStatusCodeReplacement(call.Args[4]); suggestedFix != nil {
+			suggestedFixes = append(suggestedFixes, *suggestedFix)
+		}
+	}
+	if len(suggestedFixes) == 0 {
+		return nil
+	}
+	return newHTTPConstDiagnostic(checker.Name(), call, suggestedFixes)
+}
+
+func newHTTPMethodReplacement(e ast.Expr) *analysis.SuggestedFix {
+	bt, ok := safeTypedBasicLit(e, token.STRING)
+	if !ok {
+		return nil
+	}
+	currentVal := unquoteBasicLitValue(bt)
+	key := strings.ToUpper(currentVal)
+	newVal, ok := httpMethod[key]
+	if !ok {
+		return nil
+	}
+	return newConstReplacement(bt, currentVal, newVal)
+}
+
+func newHTTPStatusCodeReplacement(e ast.Expr) *analysis.SuggestedFix {
+	bt, ok := safeTypedBasicLit(e, token.INT)
+	if !ok {
+		return nil
+	}
+	currentVal := bt.Value
+	key := strings.ToUpper(currentVal)
+	newVal, ok := httpStatusCode[key]
+	if !ok {
+		return nil
+	}
+	return newConstReplacement(bt, currentVal, newVal)
+}
+
+func newConstReplacement(bt *ast.BasicLit, currentVal, newVal string) *analysis.SuggestedFix {
+	return &analysis.SuggestedFix{
+		Message: fmt.Sprintf("Replace %q with %s", currentVal, newVal),
+		TextEdits: []analysis.TextEdit{
+			{
+				Pos:     bt.Pos(),
+				End:     bt.End(),
+				NewText: []byte(newVal),
+			},
+		},
+	}
+}
+
+func newHTTPConstDiagnostic(
+	checker string,
+	call *CallMeta,
+	suggestedFixes []analysis.SuggestedFix,
+) *analysis.Diagnostic {
+	return newAnalysisDiagnostic(checker, call, "use net/http constants instead of value", suggestedFixes)
+}
+
+func newAnalysisDiagnostic(
+	checker string,
+	rng analysis.Range,
+	msg string,
+	suggestedFixes []analysis.SuggestedFix,
+) *analysis.Diagnostic {
+	return &analysis.Diagnostic{
+		Pos:            rng.Pos(),
+		End:            rng.End(),
+		Category:       checker,
+		Message:        checker + ": " + msg,
+		SuggestedFixes: suggestedFixes,
+	}
+}
+
+func unquoteBasicLitValue(basicLit *ast.BasicLit) string {
+	s, _ := strconv.Unquote(basicLit.Value)
```

Comment:
I'm a bit surprised by the fact the current implementation ignore the error.

https://cs.opensource.google/go/go/+/go1.22.4:src/strconv/quote.go;l=382

There are a large number of use cases where it can returns an `strconv.ErrSyntax`

I would expect to handle the error in the code calling this method

Reply:
I would do this

```go
func unquoteBasicLitValue(basicLit *ast.BasicLit) (string, error) {
     return strconv.Unquote(basicLit.Value)
}
```

and when calling the code I would that
```go
	currentVal, err := unquoteBasicLitValue(bt)
	if err != nil {
	    return nil
	}
```

But then the method is somehow useless.

or to respect the pattern of the `value, ok` we have in other methods, I would do that


```go
func unquoteBasicLitValue(basicLit *ast.BasicLit) (string, bool) {
     value, err := strconv.Unquote(basicLit.Value)
     if err != nil {
          return "", false
     }
     return value, true
}
```

and when calling the code I would that
```go
	currentVal, err := unquoteBasicLitValue(bt)
	if err != nil {
	    return nil
	}
```